backend/tests/unit/controllers/posts-controller.test.js

```diff
@@ -0,0 +1,424 @@
+} from '../../../controllers/posts-controller.js';
+import Post from '../../../models/post.js';
+import { validCategories } from '../../../utils/constants.js';
+
+jest.mock('../../../models/post.js', () => ({
+  __esModule: true,
+  default: jest.fn(),
+}));
+
+const res = {
+  json: jest.fn(),
+  status: jest.fn().mockReturnThis(),
+};
+
+describe('createPostHandler', () => {
+  it('should create a new post', async () => {
+    const postObject = {
+      title: 'Test Post',
+      authorName: 'Test Author',
+      imageLink: 'https://www.forTestingPurposeOnly/my-image.jpg',
+      categories: [validCategories[0]],
+      description: 'This is a test post.',
+      isFeaturedPost: false,
+    };
+    const req = { body: postObject };
+
+    Post.mockImplementationOnce(() => ({
+      save: jest.fn().mockResolvedValueOnce(postObject),
+    }));
+
+    await createPostHandler(req, res);
+
+    expect(Post).toHaveBeenCalledTimes(1);
+    expect(Post).toHaveBeenCalledWith(postObject);
+    expect(res.status).toHaveBeenCalledWith(200);
+    expect(res.json).toHaveBeenCalledWith(postObject);
+  });
+
+  it('should handle errors - Invalid Image URL', async () => {
+    const postObject = {
+      title: 'Test Post',
+      authorName: 'Test Author',
+      // Invalid extension
+      imageLink: 'https://www.forTestingPurposeOnly/my-image.gif',
+      categories: [validCategories[0]],
+      description: 'This is a test post.',
+      isFeaturedPost: false,
+    };
+    const req = { body: postObject };
+
+    await createPostHandler(req, res);
+
+    expect(res.status).toHaveBeenCalledWith(400);
+    expect(res.json).toHaveBeenCalledWith({ message: 'Image URL must end with .jpg, .jpeg, or .png' });
+  });
+
+  it('should handle errors - Missing Form Fields', async () => {
+    const postObject = {
+      title: 'Test Post',
+      authorName: 'Test Author',
+      // Missing imageLink, categories, and description
+      isFeaturedPost: false,
+    };
+    const req = { body: postObject };
+
+    await createPostHandler(req, res);
+
+    expect(res.status).toHaveBeenCalledWith(400);
+    expect(res.json).toHaveBeenCalledWith({ message: 'All fields are required.' });
+  });
+
+  it('should handle errors - Too Many Categories', async () => {
+    const postObject = {
+      title: 'Test Post',
+      authorName: 'Test Author',
+      imageLink: 'https://www.forTestingPurposeOnly/my-image.jpg',
+      categories: [validCategories[0], validCategories[1], validCategories[2], validCategories[3]], // More than 3 categories
+      description: 'This is a test post.',
+      isFeaturedPost: false,
+    };
+    const req = { body: postObject };
+
+    await createPostHandler(req, res);
+
+    expect(res.status).toHaveBeenCalledWith(400);
+    expect(res.json).toHaveBeenCalledWith({ message: 'Please select up to three categories only.' });
+  });
+
+  it('should handle errors - Internal Server Error', async () => {
+    const postObject = {
```

Comment:
Can we refactor all the "should handle errors" to use a private function ? Please try to explore on that part if there is any refactoring possible, if there is no option we can carry on with this.